compiler/symtable_.py

```diff
@@ -0,0 +1,219 @@
+from tokens import INT
+from collections import namedtuple
+from ast.core import ScopeBlock, NodeVisitor
+from cookbook import Enum
+from collections import  OrderedDict
+
+class Symbol:
+    def __init__(self, name):
+        self.name = name
+
+    def __hash__(self):
+        return hash(self.name) + 1232344324
+
+    def __eq__(self, other):
+        return other.name == self.name
+
+
+class SymbolType(Enum):
+    pass
+
+
+VARSYMBOL = SymbolType('VarSym')
+FUNCSYMBOL = SymbolType('FuncSym')
+
+
+class EmptyStackException(Exception):
+    pass
+
+#This class should keep track of variable types in the stack, it will be needed when considering datatypes other than int.
+class StackTracker:
+    def __init__(self):
+        self._pos = 0   #  location of top of stack relative to ebp
+        self._items = []
+        self._sizes = {INT: 4}
+
+    def get_stack_pos(self):
+        return self._pos
+
+    def push(self, type_):
+        if type_ == INT:
+            self._items.append(INT)
+            self._pos -= self._items[INT]
+
+    def pop(self):
+        if self.is_empty():
+            raise EmptyStackException('Cannot pop empty stack!')
+        self._pos += self._sizes[self._items[-1]]
+        return self._items.pop()
+
+    def top_dtype(self):
+        if self.is_empty():
+            raise EmptyStackException('Empty Stack, nothing on top!')
+
+    def is_empty(self):
+        return bool(self._items)
+
+class OutOfScopeAccessException(Exception):
+    pass
+
+class ScopeObject:
+    def __init__(self,ScopeBlockNode,start_relative_to_prev):
+        self.ScopeBlockNode = ScopeBlockNode
+        self.start_relative_to_prev =  start_relative_to_prev
+        self.end_relative_to_start = 0
+    def incrementRelativeEnd(self,num):
+        self.end_relative_to_start += num
+
+    #Todo scope_object might not be active, take care for this problem later
+    def get_start_relative_to_scope(self,scope_object,scope_stack):
+        start_relative_to_scope_start = 0
+        count_on = False
+        multiple = 1
+        if scope_stack.is_active(scope_object) and scope_stack.is_active(self):
+            for temp in scope_stack:
+                if(temp == self):
+                    if (count_on == False):
+                        count_on = True
+                    else:
+                        count_on = False
+                if(temp == scope_object):
+                    if(count_on == False):
+                        count_on = True
+                        multiple = -1
+                    else:
+                        count_on = False
+                if(count_on == True):
+                    start_relative_to_scope_start +=  scope_stack.curren_scope().end_relative_to_start+4 #The +4 is because every scope block has ebp pushed
+        start_relative_to_scope_start *= multiple
```

Comment:
`+4` to compensate for ebp leads to complexity. `Special cases are not special enough to break the rules `
A better way would be to add 4 to the current scopes end before creating and pushing the new scope to the stack and to subtract it when the scope is removed in the
`Compiler.visit_ScopeStack` method